### PR TITLE
Add Oauth frontend and backend improvements

### PR DESCRIPTION
**Pull request by Boy132**
Adds oauth buttons to the filament auth.
Improves the backend and makes adding providers easier.
Adds more default providers.

**No oauth for the old auth/ old client area auth! So only admins can use oauth for now!**

Note: There is no "automatic registration" when using oauth. If you login via oauth but there is no linked user found, you will be redirected to the login page!

grafik
grafik

### CODE REVIEW

File: config/auth.php [hunk 1]
```diff
@@ -23,4 +23,126 @@
         ],
     ],
 
+    'oauth' => [
+        // Default providers
+        'facebook' => [
+            'enabled' => env('OAUTH_FACEBOOK_ENABLED'),
+            'icon' => 'tabler-brand-facebook',
+            'color' => \Filament\Support\Colors\Color::hex('#1877f2'),
```
Comment (lancepioch) [inline config/auth.php:31]: Import this at the top!

```suggestion
            'color' => Color::hex('#1877f2'),
```

File: config/auth.php [hunk 2]
```diff
@@ -23,4 +23,126 @@
...
+            'enabled' => env('OAUTH_WHMCS_ENABLED'),
+            'icon' => null,
+            'color' => \Filament\Support\Colors\Color::hex('#7bc143'),
+            'service' => [
+                'client_id' => env('OAUTH_WHMCS_CLIENT_ID'),
+                'client_secret' => env('OAUTH_WHMCS_CLIENT_SECRET'),
+                'url' => env('OAUTH_WHMCS_URL'),
+            ],
+            'provider' => \SocialiteProviders\Whmcs\Provider::class,
+        ],
```
Comment (lancepioch) [inline config/auth.php:145]:
```suggestion
```

We're gonna separate out whmcs support into plugins and separate repositories.

File: composer.json [hunk 3]
```diff
@@ -30,7 +30,10 @@
         "predis/predis": "~2.1.1",
         "ryangjchandler/blade-tabler-icons": "^2.3",
         "s1lentium/iptools": "~1.2.0",
+        "socialiteproviders/authentik": "^5.2",
         "socialiteproviders/discord": "^4.2",
+        "socialiteproviders/steam": "^4.2",
+        "socialiteproviders/whmcs": "^4.1",
```
Comment (lancepioch) [inline composer.json:36]:
```suggestion
```

We're gonna separate out whmcs support into plugins and separate repositories.

File: app/Providers/AppServiceProvider.php [hunk 4]
```diff
@@ -62,8 +61,19 @@ public function boot(Application $app): void
         Scramble::registerApi('client', ['api_path' => 'api/client', 'info' => ['version' => '1.0']])->afterOpenApiGenerated($bearerTokens);
         Scramble::registerApi('remote', ['api_path' => 'api/remote', 'info' => ['version' => '1.0']])->afterOpenApiGenerated($bearerTokens);
 
-        Event::listen(function (SocialiteWasCalled $event) {
-            $event->extendSocialite('discord', Provider::class);
+        $oauthProviders = [];
+        foreach (config('auth.oauth') as $name => $data) {
+            config()->set("services.$name", array_merge($data['service'], ['redirect' => "/auth/oauth/callback/$name"]));
```
Comment (lancepioch) [inline app/Providers/AppServiceProvider.php:66]: In the docker container we should probably cache the config on run, want to do that here or can we make an issue for that?

File: app/Filament/Resources/UserResource/Pages/EditProfile.php [hunk 5]
```diff
@@ -113,6 +115,28 @@ protected function getForms(): array
...
+                                                continue;
+                                            }
+
+                                            $unlink = array_key_exists($name, $this->getUser()->oauth);
+
+                                            $providers[] = Action::make("oauth_$name")
+                                                ->label(($unlink ? 'Unlink ' : 'Link ') . Str::title($name))
+                                                ->icon($unlink ? 'tabler-unlink' : 'tabler-link')
+                                                ->color($data['color'])
+                                                ->url('/account/oauth/' . ($unlink ? 'unlink' : 'link') .  "/$name");
```
Comment (lancepioch) [inline app/Filament/Resources/UserResource/Pages/EditProfile.php:134]: Should be able to do something like this:

```suggestion
                                                ->url(route('account.oauth.' . ($unlink ? 'unlink' : 'link'), ['driver' => $name]));
```

File: app/Filament/Pages/Auth/Login.php [hunk 6]
```diff
@@ -49,6 +53,25 @@ protected function getLoginFormComponent(): Component
...
+        foreach (config('auth.oauth') as $name => $data) {
+            if (!$data['enabled']) {
+                continue;
+            }
+
+            $actions[] = Action::make("oauth_$name")
+                ->label(Str::title($name))
+                ->icon($data['icon'])
+                ->color($data['color'])
+                ->url("/auth/oauth/redirect/$name");
```
Comment (lancepioch) [inline app/Filament/Pages/Auth/Login.php:69]:
```suggestion
                ->url(route('nameofroute'));
```

**Review by lancepioch — CHANGES_REQUESTED** [review]
Looks great so far!

File: app/Providers/AppServiceProvider.php (hunk 4 above)
Comment (Boy132) [reply app/Providers/AppServiceProvider.php:66]: Not sure what you mean by that?

**Review by lancepioch — APPROVED** [review]

**Pull request merged**